plutus-chain-index/src/Plutus/ChainIndex/Types.hs

```diff
@@ -1,20 +1,66 @@
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia        #-}
+{-# LANGUAGE NamedFieldPuns     #-}
 {-| Misc. types used in this package
 -}
 module Plutus.ChainIndex.Types(
     BlockId(..)
-    , Page
+    , Page(..)
+    , PageSize(..)
+    , pageOf
+    , Tip(..)
     ) where
 
-import qualified Data.Aeson.Extras as JSON
-import qualified Data.ByteString   as BS
+import           Data.Aeson      (FromJSON, ToJSON)
+import qualified Data.ByteString as BS
+import           Data.Default    (Default (..))
+import           Data.Set        (Set)
+import qualified Data.Set        as Set
+import           GHC.Generics    (Generic)
+import           Ledger.Bytes    (LedgerBytes (..))
+import           Ledger.Slot     (Slot)
+import           Numeric.Natural (Natural)
 
 -- | Block identifier (usually a hash)
 newtype BlockId = BlockId { getBlockId :: BS.ByteString }
-    deriving (Eq, Ord)
+    deriving stock (Eq, Ord, Generic)
+    deriving (ToJSON, FromJSON,Show) via LedgerBytes
 
-instance Show BlockId where
-    show = show . JSON.encodeByteString . getBlockId
+newtype PageSize = PageSize { getPageSize :: Natural }
+    deriving stock (Eq, Ord, Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+instance Default PageSize where
+    def = PageSize 50
 
 -- | Part of a collection
-data Page a = Page { pageSize :: Int, pageNumber :: Int, totalPages :: Int, pageItems :: [a]}
-    deriving (Eq, Ord, Show)
+data Page a = Page { pageSize :: PageSize, pageNumber :: Int, totalPages :: Int, pageItems :: [a]}
+    deriving stock (Eq, Ord, Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+-- | A page with the given size
+pageOf :: PageSize -> Set a -> Page a
+pageOf (PageSize ps) items =
+    let totalPages =
+            let (d, m) = Set.size items `divMod` ps'
+            in if m == 0 then d else d + 1
+        ps' = fromIntegral ps
+    in Page
+        { pageSize = PageSize ps
+        , pageNumber = 1
```

Comment:
When is `pageNumber` not equal to `1`?

Reply:
There isn't currently a way to request a page other than the first one (and I'm not 100% sure if this should even be possible), that's why it just returns the first page.

Maybe I should just change `Page` to `BoundedList` or something.

Reply:
The name `Page` does communicate the right idea. Just document this limitation with the `UtxoSetAtAddress` query.